src/hotspot/share/runtime/safepointMechanism.cpp

```diff
@@ -94,12 +94,31 @@ void SafepointMechanism::update_poll_values(JavaThread* thread) {
   assert(thread == Thread::current(), "Must be");
   assert(thread->thread_state() != _thread_blocked, "Must not be");
   assert(thread->thread_state() != _thread_in_native, "Must not be");
+
   for (;;) {
     bool armed = global_poll() || thread->handshake_state()->has_operation();
     uintptr_t stack_watermark = StackWatermarkSet::lowest_watermark(thread);
     uintptr_t poll_page = armed ? _poll_page_armed_value
                                 : _poll_page_disarmed_value;
     uintptr_t poll_word = compute_poll_word(armed, stack_watermark);
+    uintptr_t prev_poll_word = thread->poll_data()->get_polling_word();
+
+    if (prev_poll_word != poll_word ||
+        prev_poll_word == _poll_word_armed_value) {
+      // After updating the poll value, we allow entering new nmethods
```

Comment:
I'm a little confused about the positioning here. The comment says "after updating the poll value", but we haven't updated yet (happens below) so don't we need the fence after that point?

Reply:
The important thing is that it is called after the handshake/safepoint/stack watermark operation. The wording is just unfortunate. Maybe I should change it to "While updating the poll value" to be less confusing? Does that read better?